exporter/azuremonitorexporter/traceexporter.go

```diff
@@ -0,0 +1,660 @@
+}
+
+func formatSpanDuration(span *tracepb.Span) string {
+	startTime := toTime(span.StartTime)
+	endTime := toTime(span.EndTime)
+	return formatDuration(endTime.Sub(startTime))
+}
+
+func onAttributeStringValueExists(attributes map[string]*tracepb.AttributeValue, key string, action func(val string)) {
+	if val, exists := attributes[key]; exists {
+		action(attributeValueAsString(val))
+	}
+}
+
+func attributeValueAsString(val *tracepb.AttributeValue) string {
+	if wrapper := val.GetStringValue(); wrapper != nil {
+		return wrapper.GetValue()
+	}
+
+	return ""
+}
+
+// Transforms a wire-format Span to an AppInsights Envelope
+func spanToEnvelope(
+	instrumentationKey string,
+	span *tracepb.Span) (*contracts.Envelope, error) {
+
+	envelope := contracts.NewEnvelope()
+	envelope.Tags = make(map[string]string)
+	envelope.IKey = instrumentationKey
+	envelope.Time = toTime(span.StartTime).Format(time.RFC3339Nano)
+
+	traceIDHexString := toHex(span.TraceId)
+	envelope.Tags[contracts.OperationId] = traceIDHexString
+	envelope.Tags[contracts.OperationParentId] = formatParentChild(traceIDHexString, toHex(span.ParentSpanId))
+
+	data := contracts.NewData()
+
+	if span.Kind == tracepb.Span_SERVER {
+		requestData := spanToRequestData(span)
+		envelope.Name = requestData.EnvelopeName("")
+		envelope.Tags[contracts.OperationName] = requestData.Name
+		data.BaseData = requestData
+		data.BaseType = requestData.BaseType()
+	} else if span.Kind == tracepb.Span_CLIENT || span.Kind == tracepb.Span_SPAN_KIND_UNSPECIFIED {
+		remoteDependencyData := spanToRemoteDependencyData(span)
+		envelope.Name = remoteDependencyData.EnvelopeName("")
+		data.BaseData = remoteDependencyData
+		data.BaseType = remoteDependencyData.BaseType()
+	}
+
+	envelope.Data = data
+	sanitize(func() []string { return contracts.SanitizeTags(envelope.Tags) })
+	sanitize(func() []string { return data.Sanitize() })
+
+	return envelope, nil
+}
+
+// Transforms a wire format Span to AppInsights RequestData
+func spanToRequestData(span *tracepb.Span) *contracts.RequestData {
+	/*
+		Request type comes from a few attributes.
+
+		HTTP
+		https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md
+
+		RPC (gRPC)
+		https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-rpc.md
+
+		Database
+		https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-database.md
+	*/
+
+	// https://github.com/microsoft/ApplicationInsights-Go/blob/master/appinsights/contracts/requestdata.go
+	// Start with some reasonable default for server spans.
+	data := contracts.NewRequestData()
+	data.Id = formatTraceAndSpanAsParentChild(span)
+	data.Name = span.Name.Value
+	data.Duration = formatSpanDuration(span)
+	data.Properties = make(map[string]string)
+	data.ResponseCode = "0"
+	data.Success = true
+
+	if span.Attributes != nil && span.Attributes.AttributeMap != nil {
```

Comment:
numeric attributes can be sent as `measurements` collection

Reply:
ok